illumos-utils/src/running_zone.rs

```diff
@@ -169,13 +174,39 @@ mod zenter {
         fn ct_pr_tmpl_set_param(fd: c_int, params: c_uint) -> c_int;
         fn ct_tmpl_activate(fd: c_int) -> c_int;
         fn ct_tmpl_clear(fd: c_int) -> c_int;
+        fn ct_ctl_abandon(fd: c_int) -> c_int;
     }
 
     #[link(name = "c")]
     extern "C" {
         pub fn zone_enter(zid: zoneid_t) -> c_int;
     }
 
+    pub fn get_contract(pid: pid_t) -> std::io::Result<ctid_t> {
+        // The offset of "id_t pr_contract" in struct psinfo which is an
+        // interface documented in proc(5).
+        const CONTRACT_OFFSET: u64 = 280;
```

Comment:
This makes me a bit nervous, but I imagine this is a very stable interface? It's annoying, but we may want to write the struct out in full, with a `repr(C)` attribute, and then parse the full thing from the `psfinfo` file.

Reply:
Yes, the `/proc` interface for `psinfo` is stable. I thought this approach was less annoying than writing out the entire struct with repr(C) because it's quite large, uses a lot of different types and contains some other embedded structs. At the end of the day this is effectively equivalent.
It's a shame that libproc doesn't have some more ergonomic interfaces for this.